healthyregions/SDOHPlace
# 402 geography search

This PR implements a new "search bar", really the MapTiler [Geocoding Control](https://docs.maptiler.com/sdk-js/modules/geocoding/) and links its behavior with our existing map filter functionality.

Features of this implementation are:

- Only state, county, city, and zip codes in the US are available for search in the geocoder (streamlining usability)
- Custom style for the selected polygon with our colors
  - Zip codes are not returned from the MapTiler API with a polygon, so there is no visualization for them
  - Hacking a display of the zip code boundary built from our PMTiles layers would be kind of possible, but we wouldn't be able to make the mask in the same way, so it wouldn't be identical
  - Because the zip codes return a MultiPoint geometry, we could create a convex hull from them and use that to generate the mask. However, this would be a _very_ bad-looking shape if it were compared to our actual ZCTA layer, so I'm not sure it's really a good idea.
- Combine the `bbox` and `bboxSearch` URL params such that the latter is removed and `bbox` only appears in the URL when it is being used for a filter (this is described in https://github.com/healthyregions/SDOHPlace/issues/395#issuecomment-2498744961)
- Remove the buttons "search this area" and zoom to HI, AK, etc. as these are now redundant.

Closes #402

To test:

1. Go to the deploy preview /search
2. Type a state or county name in the new search box.
3. The map will zoom to the selected entry, the bounding box will appear in the URL, and the map filter will be enabled (this will reduce the number of search results).
4. Pan the map and you'll see the results are updated.
5. Pan to the ocean and results will go to 0
6. click the "X" to remove the search entry from the box and the bbox param will be removed and all records will be returned.

Locally:

A couple of new libraries will need to be installed with `yarn install`.

To do:

One thing I wasn't able to figure out, and it will be tricky, is how to fully clear the search box when the "CLEAR ALL" button is clicked to remove filters and reset the search. The challenging part is acquiring the control object itself. Once acquired we can run `.setQuery("")` on the control, and that will properly clear it and fire the same event as if someone has clicked the "x". @pengyin-shan If you have time to take a stab at that this week that would be great, but it's certainly something we can just address later, after the holidays, after this PR is merged.
